File: rag3d-js/src/encoders.js
// Encoders tridimensionais — espelha o HashEncoder de trirag/encoders.py.
//
// Fallback por hashing (crc32) de n-gramas de caracteres: zero dependências,
// qualquer escrita. Produz as MESMAS três representações que o Python para o
// mesmo texto (crc32 e tokenização idênticos), então os hologramas batem.
//
// dense  : hashing trick sobre n-gramas de caracteres
// sparse : palavras Unicode (CJK vira bigramas), peso 1+ln(tf)
// tokens : um vetor hasheado por palavra (MaxSim aproximado)

import { crc32str } from "./portable.js";
import { charNgrams, wordTokens } from "./textproc.js";

const hash = crc32str; // crc32 dos bytes UTF-8, sem alocar TextEncoder por chamada

function l2normalize(v) {
  let n = 0;
  for (let i = 0; i < v.length; i++) n += v[i] * v[i];
  n = Math.sqrt(n);
  if (n > 0) for (let i = 0; i < v.length; i++) v[i] /= n;
  return v;
}

export class HashEncoder {
  constructor({ denseDim = 1024, colbertDim = 128, maxTokens = 256 } = {}) {
    this.name = "hash";
    this.denseDim = denseDim;
    this.colbertDim = colbertDim;
    this.maxTokens = maxTokens;
  }

  _dense(text) {
    const v = new Float32Array(this.denseDim);
    for (const g of charNgrams(text)) {
      const h = hash(g);
      v[h % this.denseDim] += (h >>> 31) & 1 ? 1.0 : -1.0;
    }
    return l2normalize(v);
  }

  _sparse(text) {
    const tf = new Map();
    for (const w of wordTokens(text)) {
      const tid = hash("w:" + w);
      tf.set(tid, (tf.get(tid) || 0) + 1);
    }
    const out = new Map();
    for (const [t, c] of tf) out.set(t, 1.0 + Math.log(c));
    return out;
  }

  _tokenVec(word) {
    const v = new Float32Array(this.colbertDim);
    for (const g of charNgrams(word, 2, 4)) {
      const h = hash("t:" + g);
      v[h % this.colbertDim] += (h >>> 31) & 1 ? 1.0 : -1.0;
    }
    return l2normalize(v);
  }

  // devolve [{dense: Float32Array, sparse: Map, tokens: Float32Array[]}]
  encode(texts, _isQuery = false) {
    return texts.map((t) => {
      const words = wordTokens(t).slice(0, this.maxTokens);
      const tokens = words.length
        ? words.map((w) => this._tokenVec(w))
        : [new Float32Array(this.colbertDim)];
      return { dense: this._dense(t), sparse: this._sparse(t), tokens };
    });
  }
}

export function makeEncoder(cfg) {
  // JS traz só o fallback portável; BGE-M3 vive no lado Python (mesmos planos)
  return new HashEncoder({
    denseDim: cfg.denseDim,
    colbertDim: cfg.colbertDim,
    maxTokens: cfg.maxColbertTokens,
  });
}

File: rag3d-java/src/main/java/io/rag3d/core/Encoders.java
package io.rag3d.core;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Encoder tridimensional — espelha o HashEncoder de rag3d/encoders.py e
 * rag3d-js/src/encoders.js.
 *
 * <p>Fallback por hashing (crc32) de n-gramas de caracteres: zero dependências,
 * qualquer escrita. Produz as MESMAS três representações que Python/JS para o
 * mesmo texto, então os hologramas batem.
 *
 * <ul>
 *   <li>dense  — hashing trick sobre n-gramas de caracteres
 *   <li>sparse — palavras Unicode (CJK vira bigramas), peso 1+ln(tf)
 *   <li>tokens — um vetor hasheado por palavra (MaxSim aproximado)
 * </ul>
 */
public final class Encoders {

    /** As três projeções de um mesmo texto. */
    public static final class TriVec {
        public final float[] dense;
        public final Map<Long, Float> sparse;
        public final List<float[]> tokens;

        public TriVec(float[] dense, Map<Long, Float> sparse, List<float[]> tokens) {
            this.dense = dense;
            this.sparse = sparse;
            this.tokens = tokens;
        }
    }

    public static final class HashEncoder {
        public final String name = "hash";
        private final int denseDim;
        private final int colbertDim;
        private final int maxTokens;

        public HashEncoder(int denseDim, int colbertDim, int maxTokens) {
            this.denseDim = denseDim;
            this.colbertDim = colbertDim;
            this.maxTokens = maxTokens;
        }

        public HashEncoder() {
            this(1024, 128, 256);
        }

        private static float[] l2(float[] v) {
            double n = 0;
            for (float x : v) n += (double) x * x;
            n = Math.sqrt(n);
            if (n > 0) for (int i = 0; i < v.length; i++) v[i] = (float) (v[i] / n);
            return v;
        }

        private float[] dense(String text) {
            float[] v = new float[denseDim];
            for (String g : TextProc.charNgrams(text)) {
                long h = Portable.crc32(g);
                v[(int) (h % denseDim)] += ((h >> 31) & 1L) != 0 ? 1.0f : -1.0f;
            }
            return l2(v);
        }

        private Map<Long, Float> sparse(String text) {
            Map<Long, Integer> tf = new LinkedHashMap<>();
            for (String w : TextProc.wordTokens(text)) {
                long tid = Portable.crc32("w:" + w);
                tf.merge(tid, 1, Integer::sum);
            }
            Map<Long, Float> out = new LinkedHashMap<>();
            for (Map.Entry<Long, Integer> e : tf.entrySet()) {
                out.put(e.getKey(), (float) (1.0 + Math.log(e.getValue())));
            }
            return out;
        }

        private float[] tokenVec(String word) {
            float[] v = new float[colbertDim];
            for (String g : TextProc.charNgrams(word, 2, 4)) {
                long h = Portable.crc32("t:" + g);
                v[(int) (h % colbertDim)] += ((h >> 31) & 1L) != 0 ? 1.0f : -1.0f;
            }
            return l2(v);
        }

        public TriVec encode(String text) {
            List<String> words = TextProc.wordTokens(text);
            if (words.size() > maxTokens) words = words.subList(0, maxTokens);
            List<float[]> tokens = new ArrayList<>();
            if (words.isEmpty()) {
                tokens.add(new float[colbertDim]);
            } else {
                for (String w : words) tokens.add(tokenVec(w));
            }
            return new TriVec(dense(text), sparse(text), tokens);
        }
    }

    private Encoders() {}
}
